text/0000-cargo-prepublish.md

```diff
@@ -0,0 +1,395 @@
+specify the source that's being augmented. Cargo will know ahead of time one
+identifier, literally `crates-io`, but otherwise this field will currently be
+interpreted as a URL of a source. The name `crates-io` will correspond to the
+crates.io index, and other urls, such as git repositories, may also be specified
+for augmentation. Eventually it's intended we'll grow support for multiple
+registries here with their own identifiers, but for now just literally
+`crates-io` and other URLs are allowed.
+
+## Examples
+
+It's easiest to see how the feature works by looking at a few examples.
+
+Let's imagine that `xml-rs` is currently at version `0.9.1` on crates.io, and we
+have the following dependency setup:
+
+- Crate `foo` lists dependency `xml-rs = "0.9.0"`
+- Crate `bar` lists dependency `xml-rs = "0.9.1"`
+- Crate `baz` lists dependency `xml-rs = "0.8.0"`
+- Crate `servo` has `foo`, `bar` and `baz` as dependencies.
+
+With this setup, the dependency graph for Servo will contain *two* versions of
+`xml-rs`: `0.9.1` and `0.8.0`. That's because minor versions are coalesced;
+`0.9.1` is considered a minor release against `0.9.0`, while `0.9.0` and `0.8.0`
+are incompatible.
+
+### Scenario: augmenting with a bugfix
+
+Let's say that while developing `foo` we've got a lock file pointing to `xml-rs`
+`0.8.0`, and there's an `0.8.0` branch of `xml-rs` that hasn't been touched
+since it was published. We then find a bug in the 0.8.0 publication of `xml-rs`
+which we'd like to fix.
+
+First we'll check out `foo` locally and implement what we believe is a fix for
+this bug, and next, we change `Cargo.toml` for `foo`:
+
+```toml
+[augment.crates-io]
+xml-rs = { path = "../xml-rs" }
+```
+
+When compiling `foo`, Cargo will resolve the `xml-rs` dependency to `0.8.0`,
+as it did before, but that version's been replaced with our local copy. The
+local path dependency, which has version 0.8.0, takes precedence over the
+version found in the registry.
+
+Once we've confirmed a fix bug we then continue to run tests in `xml-rs` itself,
+and then we'll send a PR to the main `xml-rs` repo. This then leads us to the
+next section where a new version of `xml-rs` comes into play!
+
+### Scenario: prepublishing a new minor version
+
+Now, suppose that `foo` needs some changes to `xml-rs`, but we want to check
+that all of Servo compiles before pushing the changes through.
+
+First, we change `Cargo.toml` for `foo`:
+
+```toml
+[augment.crates-io]
+xml-rs = { git = "https://github.com/aturon/xml-rs", branch = "0.9.2" }
+
+[dependencies]
+xml-rs = "0.9.2"
+```
+
+For `servo`, we also need to record the prepublication, but don't need to modify
+or introduce any `xml-rs` dependencies; it's enough to be using the fork of
+`foo`, which we would be anyway:
+
+```toml
+[augment.crates-io]
+xml-rs = { git = "https://github.com/aturon/xml-rs", branch = "0.9.2" }
+
+[dependencies]
+foo = { git = "https://github.com/aturon/foo", branch = "fix-xml" }
+```
```

Comment:
I'm not sure I understand this example-- why would we use `augment` when we're working on `foo` that needs changes to `xml-rs`, but then when we're working on `servo` that needs changes to `foo`, we change the `foo` dependency line rather than using `augment` with foo?

That is, why wouldn't servo's Cargo.toml look like this?

```
[augment.crates-io]
xml-rs = { git = "https://github.com/aturon/xml-rs", branch = "0.9.2" }
foo = { git = "https://github.com/aturon/foo", branch = "fix-xml" }

[dependencies]
foo = "*" # whatever servo's foo version requirement was before we started all this
```

Also it says "For `servo`, we also need to record the prepublication, but don't need to modify or introduce any `xml-rs` dependencies;" but I would describe adding `augment` as "modifying an xml-rs dependency"... could this be clarified as "For `servo`, we also need to record the prepublication, but don't need to modify or introduce any direct `xml-rs` dependencies"?